src/helperFunctions/formHelper.ts

```diff
@@ -21,8 +20,249 @@ import {
   FormData,
   GenericStringArray,
 } from "../types/express";
+import { licences } from "../mockData/licences";
 import { replace } from "./checkhelper";
 
+export const updateStepsStatus = (
+  currentStep: string,
+  stepValue: StepValue,
+  formdata: FormData,
+  returnToStart: boolean,
+) => {
+  const completedSections = new Set();
+  // Group up the steps so we can work out which sections have been completed later
+  const purposeSteps = [
+    "data-type",
+    "data-subjects",
+    "project-aims",
+    "data-required",
+    "benefits",
+    "data-access",
+    "other-orgs",
+    "impact",
+    "date",
+  ];
+
+  const legalSteps = [
+    "legal-power",
+    "legal-power-advice",
+    "legal-gateway",
+    "legal-gateway-advice",
+  ];
+
+  const dataProtectionSteps = [
+    "lawful-basis-personal",
+    "lawful-basis-special",
+    "lawful-basis-special-public-interest",
+    "data-travel",
+    "data-travel-location",
+    "role",
+  ];
+
+  const securitySteps = ["delivery", "format", "disposal"];
+
+  // If "Save and return" was clicked, set to "IN PROGRESS" if needed and return
+  if (returnToStart) {
+    if (formdata.steps[currentStep].status !== "COMPLETED") {
+      formdata.steps[currentStep].status = "IN PROGRESS";
+    }
+    return;
+  } else {
+    // If "Save and Continue" was clicked, set this step to "COMPLETED"
+    formdata.steps[currentStep].status = "COMPLETED";
+  }
+
+  if (currentStep === "data-type") {
+    const notRequiredSteps = new Set<string>();
+    const notStartedSteps = new Set<string>();
+    const val = stepValue as DataTypeStep;
+    // If personal is not checked then lawful-basis-personal is not required
+    if (!val.personal.checked) {
+      notRequiredSteps.add("lawful-basis-personal");
+    } else {
+      notStartedSteps.add("data-subjects");
+      notStartedSteps.add("lawful-basis-personal");
+      notStartedSteps.add("role");
+    }
+    // If special is not checked then lawful-basis-special is not required
+    if (!val.special.checked) {
+      notRequiredSteps.add("lawful-basis-special");
+      notRequiredSteps.add("lawful-basis-special-public-interest");
+    } else {
+      notStartedSteps.add("data-subjects");
+      notStartedSteps.add("lawful-basis-special");
+      notStartedSteps.add("role");
+    }
+    // If none is checked, a few other steps are not required
+    if (val.none.checked) {
+      [
+        "data-subjects",
+        "lawful-basis-personal",
+        "lawful-basis-special",
+        "lawful-basis-special-public-interest",
+        "role",
+      ].forEach((s) => notRequiredSteps.add(s));
+    }
+
+    // Set everything that's not required to NOT REQUIRED
+    for (const s of notRequiredSteps) {
+      if (Object.prototype.hasOwnProperty.call(formdata.steps, s)) {
```

Comment:
@kjoshi I had to put these checks in for the tests as I dont have full form mock data. Can you think of anything adverse by doing this? The other option would be to ensure the test has enough mockdata that this doesnt fail.